apps/beeswax/src/beeswax/server/hive_server2_lib.py

```diff
@@ -617,8 +619,64 @@ def open_session(self, user):
     return session
 
 
-  def call(self, fn, req, status=TStatusCode.SUCCESS_STATUS):
-    session = Session.objects.get_session(self.user, self.query_server['server_name'])
+  def call(self, fn, req, status=TStatusCode.SUCCESS_STATUS,
+           withMultipleSession=False):
+    (res, session) = self.call_return_result_and_session(fn, req, status, withMultipleSession)
+    return res
+
+
+  def call_return_result_and_session(self, fn, req, status=TStatusCode.SUCCESS_STATUS,
+                                     withMultipleSession=False):
+
+    n_sessions = conf.MAX_NUMBER_OF_SESSIONS.get()
+
+    # When a single session is allowed, avoid multiple session logic
+    if n_sessions == 1:
+      withMultipleSession = False
+
+    session = None
+
+    if not withMultipleSession:
+
+      # Default behaviour: get one session
+      session = Session.objects.get_session(self.user, self.query_server['server_name'])
+
+    else:
+
+      # Get 2 + n_sessions sessions and filter out the busy ones
```

Comment:
Curious: is the algo like this in order to avoid adding a Foreign Key from a session to a query?
(that way we could get the last 10 sessions and their queries and quickly check the status of the query/session)

Isn't there a way to know if a session is busy directly with a certain Thrift call? (would even be simpler)